### PR TITLE
feat(core): optimized asynchronous handling of graph transitions

### LINKED ISSUE #1143: [Bug]: Transitions are failing in Salesforce Lightning Web Components
### Contact Details

steffen.baumann@de.ibm.com

### What happened?

Carbon Charts are attempted to be hosted within a Salesforce Lightning Web Component (LWC). Adding Carbon Charts as third party library works fine. Embedding cahrts seem to work in general as well. Though, transitions are causing issues:
### SimpleBar Chart:
(1.) Mouseover on a bar causes a console error. No tooltip with data is displayed
(2.) Mouseout on a bar causes another console error.
(3.) Hovering the legend items causes several console log errors. Fading out of bars doesn't work

### Radar Chart
(4.) Loading the diagram fails due to some transition error. The legend and title are visible.

Reviews of the underlying code suggests that using named transitions like this:
```
this.parent
	.selectAll('path.bar')
	.transition(
		this.services.transitions.getTransition(
			'legend-hover-simple-bar'
		)
	)
	.attr('opacity', (d) =>
		d[groupMapsTo] !== hoveredElement.datum()['name'] ? 0.3 : 1
	);
```
are causing errors in the utilized d3.js library. It appears that using the d3-transition & d3-selection modules, in order to create a new transition as part of the factory pattern, generates it on the document root element.

**Salesforce follows a very strict Shadow-DOM approach on their Web Components.** Therefore, the generation on the root element seems problematic. Interestingly, other transitions in Carbon Charts seem to work just fine (e.g. when clicking a dataset in the legend). This suggests, different approaches are taken on other places:
```
barsEnter
	.merge(bars)
	.classed('bar', true)
	.attr('width', this.getBarWidth.bind(this))
	.transition()
	.call((t) =>
		this.services.transitions.setupTransition({
			transition: t,
			name: 'bar-update-enter',
			animate,
		})
	)
```

Verifying the different approaches with a plain d3.js library confirmed, the `setupTransition` approach is working whereas `getTransition` fails in a Salesforce LWC.

This said, moving to `setupTransition` instead `getTransition` for all existing transitions should resolve the issue and allows embedding Carbon Charts in Salesforce LWC.

### Version

0.47.0

### Data & options used

```js
Utilized Sample data & options from published demos:

a) Simple Bar Chart
https://carbon-design-system.github.io/carbon-charts/?path=/story/simple-charts-bar-simple--vertical-simple-bar-discrete

b) Radar Chart:
https://carbon-design-system.github.io/carbon-charts/?path=/story/simple-charts-radar--radar
```

### Relevant log output

```js
1. Mouseover on bar:
bundle.min.js:15 Uncaught Error: transition 6 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at SVGPathElement.<anonymous> (bundle.min.js:15)
    at SVGPathElement.<anonymous> (bundle.min.js:15)
    at SVGPathElement.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
---

2. Mouseout on bar
bundle.min.js:15 Uncaught Error: transition 7 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at SVGPathElement.<anonymous> (bundle.min.js:15)
    at SVGPathElement.<anonymous> (bundle.min.js:15)
    at SVGPathElement.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
---

3. Hovering through legend item
Uncaught Error: transition 8 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at DocumentFragment.e.handleLegendOnHover (bundle.min.js:15)
    at DocumentFragment.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
    at e.dispatchEvent (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
Uncaught Error: transition 9 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at DocumentFragment.e.handleLegendMouseOut (bundle.min.js:15)
    at DocumentFragment.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
    at e.dispatchEvent (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
Uncaught Error: transition 10 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at DocumentFragment.e.handleLegendOnHover (bundle.min.js:15)
    at DocumentFragment.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
    at e.dispatchEvent (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
Uncaught Error: transition 11 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at DocumentFragment.e.handleLegendMouseOut (bundle.min.js:15)
    at DocumentFragment.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
    at e.dispatchEvent (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
Uncaught Error: transition 12 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at DocumentFragment.e.handleLegendOnHover (bundle.min.js:15)
    at DocumentFragment.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
    at e.dispatchEvent (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
Uncaught Error: transition 13 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at DocumentFragment.e.handleLegendMouseOut (bundle.min.js:15)
    at DocumentFragment.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)
    at e.dispatchEvent (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
    at HTMLDivElement.listener.$$lwcEventWrapper$$ (webruntime-amd-shim.js:4068)

4. Radar loading issue due transition
Error: transition 1 not found
    at Wp (bundle.min.js:15)
    at Pa.La.transition (bundle.min.js:15)
    at bundle.min.js:15
    at Pa.call (bundle.min.js:15)
    at bundle.min.js:15
    at Pa.join (bundle.min.js:15)
    at e.render (bundle.min.js:15)
    at bundle.min.js:15
    at Array.forEach (<anonymous>)
    at HTMLDivElement.<anonymous> (bundle.min.js:15)
```

### Codesandbox example

_No response_

### PR DESCRIPTION
**Pull request by theiliad**
Closes #1143

### CODE REVIEW

File: packages/core/src/components/component.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ import { RenderTypes } from '../interfaces';
 import { Tools } from '../tools';
 
 // D3 Imports
-import { select } from 'd3-selection';
+import { select, Selection } from 'd3-selection';
```
Comment (Akshat55) [inline packages/core/src/components/component.ts:8]: Selection interface is imported, but not used?

File: packages/core/src/components/graphs/bar-grouped.ts [hunk 2]
```diff
@@ -187,41 +192,46 @@ export class GroupedBar extends Bar {
 
 		const { groupMapsTo } = this.getOptions().data;
```
Comment (Akshat55) [inline packages/core/src/components/graphs/bar-grouped.ts:193]: Destructuring should be inside if condition

File: packages/core/src/components/graphs/bar-simple.ts [hunk 3]
```diff
@@ -135,51 +141,55 @@ export class SimpleBar extends Bar {
...
-		this.addEventListeners();
+			.attr('aria-label', (d) => d.value)
+			.on('end', () => {
+				this.componentContainer.classed('updating', false);
+			});
 	}
 
 	handleLegendOnHover = (event: CustomEvent) => {
 		const { hoveredElement } = event.detail;
 		const { groupMapsTo } = this.getOptions().data;
```
Comment (Akshat55) [inline packages/core/src/components/graphs/bar-simple.ts:152]: inside condition

File: packages/core/src/components/graphs/bar-stacked.ts [hunk 4]
```diff
@@ -169,43 +174,48 @@ export class StackedBar extends Bar {
 
 		const { groupMapsTo } = this.model.getOptions().data;
```
Comment (Akshat55) [inline packages/core/src/components/graphs/bar-stacked.ts:175]: Line 173 - 175 should be  nested within the if condition similar to area and area-stacked

**Review by Akshat55 — CHANGES_REQUESTED** [review]

**Review by Akshat55 — APPROVED** [review]

File: packages/core/src/components/graphs/wordcloud.ts [hunk 5]
```diff
@@ -32,9 +32,12 @@ export class WordCloud extends Component {
 
 	render(animate = true) {
 		const self = this;
-		const svg = this.getComponentContainer()
+
```
Comment (Akshat55) [inline packages/core/src/components/graphs/wordcloud.ts:35]: Add a destroy function and call it here, seems like even after destroy is added the function is never called. 
```suggestion
this.destory();
```

In the destroy function set all (click, mouseover, mouseout) event listeners to null 

```suggestion
	destroy() {
		this.parent.selectAll('text.word')
			.on('mouseover', null)
			.on('mousemove', null)
			.on('click', null)
			.on('mouseout', null);
	}
```

File: packages/core/src/components/graphs/wordcloud.ts [hunk 6]
```diff
@@ -106,18 +118,20 @@ export class WordCloud extends Component {
...
+						name: 'wordcloud-text-update-enter',
+						animate,
+					})
 				)
 				.attr('transform', (d) => `translate(${d.x}, ${d.y})`)
-				.attr('opacity', 1);
+				.attr('opacity', 1)
+				.on('end', () => {
+					self.componentContainer.classed('updating', false);
+				});
```
Comment (Akshat55) [inline packages/core/src/components/graphs/wordcloud.ts:133]: Class updating is being removed (By setting to false) from the element n number of times (In demo example, 42).

Switch approach to async/await or promises.

```suggestion
				.end();
				
				//Assign merged text to promise
				promise.then(() => {
				  // Give users the option to hover and select legend after everything loads
				  self.addEventListeners(mergedText);
				  self.componentContainer.classed('updating', false); 
				});
```

**Review by Akshat55 — CHANGES_REQUESTED** [review]

**Comment by theiliad** [comment]
Closed by #1169

**Pull request closed without merging**